### PR TITLE
DOMA-3874 marketing messages with custom title and content being sent to idList

### PR DESCRIPTION
**Pull request by akarjakin**
- added schema
- added new message type
- added message templates
- added migration
- added schema tests
- added task
- added task & helpers tests

### CODE REVIEW

File: apps/condo/domains/common/utils/testSchema/index.js [hunk 1]
```diff
@@ -112,6 +112,7 @@ const expectToThrowAccessDeniedErrorToResult = async (testFunc) => {
 const expectToThrowAuthenticationError = async (testFunc, path) => {
     if (!path) throw new Error('expectToThrowAccessDeniedError(): no path argument')
     await catchErrorFrom(testFunc, (caught) => {
+        console.log('caught:', caught)
```
Comment (pahaz) [inline apps/condo/domains/common/utils/testSchema/index.js:115]: !!

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+const operationForbiddenValidator = ({ fieldPath, addFieldValidationError, operation }) => {
+    if (operation === 'update') addFieldValidationError(`${OPERATION_FORBIDDEN}] Updating ${fieldPath} is forbidden.`)
+}
+
+const MarketingMessage = new GQLListSchema('MarketingMessage', {
+    schemaDoc: 'Marketing message batch. Adding record here will cause sending message to all users listed in idList via transport, selected in transportType.',
+    fields: {
+
+        transportType: {
+            schemaDoc: 'Transport type to send notifications by.',
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:35]:
```suggestion
            schemaDoc: 'Transport type to send message by',
```

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            type: Select,
+            options: MARKETING_MESSAGE_TRANSPORT_TYPES,
+            isRequired: true,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        title: {
+            schemaDoc: 'Title for message to send.',
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:45]:
```suggestion
            schemaDoc: 'Title for a message to send',
```

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            type: Text,
+            isRequired: true,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        message: {
+            schemaDoc: 'Message to send.',
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:54]:
```suggestion
            schemaDoc: 'Message to send',
```

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        deepLink: {
+            schemaDoc: 'Used for push notifications only. Useful for moving user to proper position within mobile apps.',
+            type: Text,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        idList: {
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:70]: probably it should be: ID list (for PUSH/EMAIL/SMS) / email list (for Email) / phone list (for SMS)

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    access: {
+        read: access.canReadMarketingMessages,
+        create: access.canManageMarketingMessages,
+        update: access.canManageMarketingMessages,
+        delete: false,
+        auth: true,
+    },
+    hooks: {
+        resolveInput: async ({ operation, resolvedData }) => {
+            if (operation === 'create')  resolvedData.status = MARKETING_MESSAGE_CREATED_STATUS
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:115]:
```suggestion
            if (operation === 'create') resolvedData.status = MARKETING_MESSAGE_CREATED_STATUS
```

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+            return resolvedData
+        },
+        afterChange: async ({ operation, updatedItem }) => {
+            if (operation === 'create')  await sendMarketingMessagesTask.delay(updatedItem.id)
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:120]:
```suggestion
            if (operation === 'create') await sendMarketingMessagesTask.delay(updatedItem.id)
```

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+const operationForbiddenValidator = ({ fieldPath, addFieldValidationError, operation }) => {
+    if (operation === 'update') addFieldValidationError(`${OPERATION_FORBIDDEN}] Updating ${fieldPath} is forbidden.`)
+}
+
+const MarketingMessage = new GQLListSchema('MarketingMessage', {
+    schemaDoc: 'Marketing message batch. Adding record here will cause sending message to all users listed in idList via transport, selected in transportType.',
+    fields: {
+
+        transportType: {
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:34]: We don't need to set the transport. It should not be the part of the model

File: apps/condo/domains/notification/schema/MarketingMessage.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            schemaDoc: 'Message to send.',
+            type: Text,
+            isRequired: true,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        deepLink: {
+            schemaDoc: 'Used for push notifications only. Useful for moving user to proper position within mobile apps.',
```
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:63]:
```suggestion
            schemaDoc: 'Used for push notifications only. Useful for moving a user to proper position within mobile apps',
```

File: apps/condo/domains/notification/schema/MarketingMessage.js (hunk 7 above)
Comment (pahaz) [inline apps/condo/domains/notification/schema/MarketingMessage.js:120]: You should not use `Task` in function name

File: apps/condo/domains/notification/tasks/sendMarketingMessagesTask.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    return { user: { id: user.id } }
+}
+
+const getUniqKey = (batchId, date) => `${batchId}:${date}`
+
+async function sendMarketingMessages (batchId) {
+    const { keystone: context } = await getSchemaCtx('MarketingMessage')
+    const batch = await MarketingMessage.getOne(context, { id: batchId })
+
+    if (batch.id !== batchId) throw new Error('get batch by id has wrong result')
```
Comment (pahaz) [inline apps/condo/domains/notification/tasks/sendMarketingMessagesTask.js:38]: looks strange

File: apps/condo/domains/notification/tasks.js [hunk 11]
```diff
@@ -3,10 +3,12 @@ const isEmpty = require('lodash/isEmpty')
 const conf = require('@condo/config')
 const { createTask } = require('@condo/keystone/tasks')
 const { getSchemaCtx } = require('@condo/keystone/schema')
+const { getLogger } = require('@condo/keystone/logging')
```
Comment (pahaz) [inline apps/condo/domains/notification/tasks.js:6]: if you create a `tasks` folder you should move `tasks.js` inside the folder

File: apps/condo/domains/notification/constants/constants.js [hunk 12]
```diff
@@ -338,6 +340,16 @@ const MESSAGE_META = {
             url: { defaultValue: '', required: true },
         },
     },
+    [CUSTOM_CONTENT_MESSAGE_TYPE]: {
+        dv: { required: true },
+        title: { required: true },
+        body: { required: true },
+        data: {
+            userId: { required: false },
```
Comment (pahaz) [inline apps/condo/domains/notification/constants/constants.js:348]: what about userName ?

File: apps/condo/domains/notification/tasks/sendMarketingMessagesTask.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }
+
+    await MarketingMessage.update(context, batch.id, {
+        ...baseAttrs,
+        status: MARKETING_MESSAGE_DONE_STATUS,
+    })
+}
+
+const sendMarketingMessagesTask = createTask('sendMarketingMessages', sendMarketingMessages, { priority: 3 })
```
Comment (pahaz) [inline apps/condo/domains/notification/tasks/sendMarketingMessagesTask.js:108]: don't use priority here.

File: apps/condo/domains/notification/utils/clientSchema/MarketingMessage.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+import {
+    MarketingMessage,
+    MarketingMessageCreateInput,
+    MarketingMessageUpdateInput,
+    QueryAllMarketingMessagesArgs,
+} from '@app/condo/schema'
+import { generateReactHooks } from '@condo/domains/common/utils/codegeneration/generate.hooks'
+import { MarketingMessage as MarketingMessageGQL } from '@condo/domains/notification/gql'
+
+// TODO(codegen): write utils like convertToFormState and formValuesProcessor if needed, otherwise delete this TODO
```
Comment (pahaz) [inline apps/condo/domains/notification/utils/clientSchema/MarketingMessage.ts:14]: ??

File: apps/condo/domains/notification/utils/testSchema/index.js [hunk 15]
```diff
@@ -31,6 +31,8 @@ const { SET_MESSAGE_STATUS_MUTATION } = require('@condo/domains/notification/gql
 const { MessageUserBlackList: MessageUserBlackListGQL } = require('@condo/domains/notification/gql')
 const { MessageOrganizationBlackList: MessageOrganizationBlackListGQL } = require('@condo/domains/notification/gql')
 const { MessageOrganizationWhiteList: MessageOrganizationWhiteListGQL } = require('@condo/domains/notification/gql')
+const { MarketingMessage: MarketingMessageGQL } = require('@condo/domains/notification/gql')
+const { MARKETING_MESSAGE_TRANSPORT_TYPES } = require('../../constants/constants')
```
Comment (pahaz) [inline apps/condo/domains/notification/utils/testSchema/index.js:35]: import

**Comment by akarjakin** [comment]
moved there https://github.com/open-condo-software/condo/pull/2280

**Pull request closed without merging**